src/main/java/org/apache/flume/sink/pulsar/PulsarSink.java

```diff
@@ -203,11 +208,28 @@ public Status process() throws EventDeliveryException {
                 } else {
                     counter.incrementBatchCompleteCount();
                 }
-                producer.send(serializeEvent(event, useAvroEventFormat));
+                TypedMessageBuilder<byte[]> newMessage = producer.newMessage();
+                if (event.getHeaders() != null) {
+                    if (event.getHeaders().get("key") != null) {
+                        newMessage = newMessage.key(event.getHeaders().get("key"));
+                    }
+                    newMessage.value(serializeEvent(event, useAvroEventFormat)).properties(event.getHeaders());
+                }
+                if (syncMode) {
+                    newMessage.send();
+                } else {
+                    newMessage.sendAsync().thenAccept(msgId -> {
+                        log.info("Success send message in asyncMode {}", msgId);
+                    });
+                }
+            }
+            if (syncMode) {
```

Comment:
@tuteng if `syncMode` is enabled, you don't need to flush. `send()` only returns when a message is flushed and persistent.

I don't think `flushAsync` should be used here. based on my understanding on flume transaction, you have to wait until all messages are flushed and persisted before committing the transaction.

Reply:
good, i see